File: sdk/java-sdk/src/main/java/com/evernym/verity/sdk/protocols/presentproof/v1_0/PresentProofV1_0.java
package com.evernym.verity.sdk.protocols.presentproof.v1_0;

import com.evernym.verity.sdk.exceptions.VerityException;
import com.evernym.verity.sdk.protocols.MessageFamily;
import com.evernym.verity.sdk.protocols.Protocol;
import com.evernym.verity.sdk.protocols.presentproof.common.Attribute;
import com.evernym.verity.sdk.protocols.presentproof.common.Predicate;
import com.evernym.verity.sdk.protocols.presentproof.common.ProposedAttribute;
import com.evernym.verity.sdk.protocols.presentproof.common.ProposedPredicate;
import com.evernym.verity.sdk.protocols.presentproof.common.Restriction;
import com.evernym.verity.sdk.utils.Context;
import com.evernym.verity.sdk.utils.Util;
import org.json.JSONObject;

import java.io.IOException;

/**
 * An interface for controlling a 1.0 PresentProof protocol.
 *
 * @see <a href="https://github.com/hyperledger/aries-rfcs/tree/4fae574c03f9f1013db30bf2c0c676b1122f7149/features/0037-present-proof" target="_blank" rel="noopener noreferrer">Aries 0037: Present Proof Protocol 1.0</a>
 */
@SuppressWarnings("CPD-START")
public interface PresentProofV1_0 extends Protocol {
    /**
     * The qualifier for the message family. Uses the community qualifier.
     */
    String QUALIFIER = Util.COMMUNITY_MSG_QUALIFIER;
    /**
     * The name for the message family.
     */
    String FAMILY = "present-proof";
    /**
     * The version for the message family.
     */
    String VERSION = "1.0";


    /**
     * @see MessageFamily#qualifier()
     */
    default String qualifier() {return QUALIFIER;}
    /**
     * @see MessageFamily#family()
     */
    default String family() { return FAMILY;}
    /**
     * @see MessageFamily#version()
     */
    default String version() {return VERSION;}

    /**
     * Creates an attribute restriction for a presentation of proof
     * @param name the attribute name to apply the restriction
     * @param restrictions an array restrictions to be used
     * @return An Attribute with the given name and restrictions
     */
    static Attribute attribute(String name, Restriction... restrictions) {
        return new Attribute(name, restrictions);
    }

    /**
     * Creates a predicate restriction for a presentation of proof. Indy Anoncreds only supports, so the
     * value should be expressed as a greater than predicate.
     *
     * @param name the attribute name to apply the restriction
     * @param value the value the attribute must be greater than
     * @param restrictions an array restrictions to be used
     * @return A Predicate with the given name, value and restrictions
     */
    static Predicate predicate(String name, int value, Restriction... restrictions) {
        return new Predicate(name, value, restrictions);
    }

    /**
     * Creates a proposed attribute value for a presentation of proof.
     *
     * @param name the attribute name
     * @param credDefId cred def id of credential being used for this attribute. If null, than self attested.
     * @param value the value of the attribute. If null, then value want be revealed.
     * @return A ProposedAttribute with the given name, credDefId and value
     */
    static ProposedAttribute proposedAttribute(String name, String credDefId, String value) {
        return new ProposedAttribute(name, credDefId, value);
    }

    /**
     * Creates a proposed predicate for a presentation of proof. Indy Anoncreds only supports, so the
     * value should be expressed as a greater than predicate.
     *
     * @param name the predicate name
     * @param credDefId cred def id of credential being used for proposed predicate.
     * @param treshold the value the predicate must be greater than
     * @return A ProposedPredicate with the given name, credDefId, predicate type and treshold
     */
    static ProposedPredicate proposedPredicate(String name, String credDefId, int treshold) {
        return new ProposedPredicate(name, credDefId, treshold);
    }

    /**
     * Directs verity-application to request a presentation of proof.
     *
     * @param context an instance of the Context object initialized to a verity-application agent
     * @throws IOException when the HTTP library fails to post to the agency endpoint
     * @throws VerityException when wallet operations fails or given invalid context
     */
    void request(Context context) throws IOException, VerityException;

    /**
     * Creates the control message without packaging and sending it.
     * @param context an instance of the Context object initialized to a verity-application agent
     * @return the constructed message (JSON object)
     * @throws VerityException when given invalid context
     *
     * @see #request
     */
    JSONObject requestMsg(Context context) throws VerityException;

    /**
     * Creates and packages message without sending it.
     * @param context an instance of the Context object initialized to a verity-application agent
     * @return the byte array ready for transport
     * @throws VerityException when wallet operations fails or given invalid context
     *
     * @see #request
     */
    byte[] requestMsgPacked(Context context) throws VerityException;

    /**
     * Directs verity-application to propose a presentation of proof.
     *
     * @param context an instance of the Context object initialized to a verity-application agent
     * @throws IOException when the HTTP library fails to post to the agency endpoint
     * @throws VerityException when wallet operations fails or given invalid context
     */
    void propose(Context context) throws IOException, VerityException;

    /**
     * Creates the control message without packaging and sending it.
     * @param context an instance of the Context object initialized to a verity-application agent
     * @return the constructed message (JSON object)
     * @throws VerityException when given invalid context
     *
     * @see #propose
     */
    JSONObject proposeMsg(Context context) throws VerityException;

    /**
     * Creates and packages message without sending it.
     * @param context an instance of the Context object initialized to a verity-application agent
     * @return the byte array ready for transport
     * @throws VerityException when wallet operations fails or given invalid context
     *
     * @see #propose
     */
    byte[] proposeMsgPacked(Context context) throws VerityException;

    /**
     * Directs verity-application to accept the request to present proof.
     *
     * @param context an instance of the Context object initialized to a verity-application agent
     * @throws IOException when the HTTP library fails to post to the agency endpoint
     * @throws VerityException when wallet operations fails or given invalid context
     */
    void acceptRequest(Context context) throws IOException, VerityException;

    /**
     * Creates the control message without packaging and sending it.
     * @param context an instance of the Context object initialized to a verity-application agent
     * @return the constructed message (JSON object)
     * @throws VerityException when given invalid context
     *
     * @see #acceptRequest
     */
    JSONObject acceptRequestMsg(Context context) throws VerityException;

    /**
     * Creates and packages message without sending it.
     * @param context an instance of the Context object initialized to a verity-application agent
     * @return the byte array ready for transport
     * @throws VerityException when wallet operations fails or given invalid context
     *
     * @see #acceptRequest
     */
    byte[] acceptRequestMsgPacked(Context context) throws VerityException;

    /**
     * Directs verity-application to accept the proposal for present proof.
     * verity-application will send presentation request based on the proposal.
     *
     * @param context an instance of the Context object initialized to a verity-application agent
     * @throws IOException when the HTTP library fails to post to the agency endpoint
     * @throws VerityException when wallet operations fails or given invalid context
     */
    void acceptProposal(Context context) throws IOException, VerityException;

    /**
     * Creates the control message without packaging and sending it.
     * @param context an instance of the Context object initialized to a verity-application agent
     * @return the constructed message (JSON object)
     * @throws VerityException when given invalid context
     *
     * @see #acceptProposal
     */
    JSONObject acceptProposalMsg(Context context) throws VerityException;

    /**
     * Creates and packages message without sending it.
     * @param context an instance of the Context object initialized to a verity-application agent
     * @return the byte array ready for transport
     * @throws VerityException when wallet operations fails or given invalid context
     *
     * @see #acceptProposal
     */
    byte[] acceptProposalMsgPacked(Context context) throws VerityException;


    /**
     * Directs verity-application to reject this presentation proof protocol
     *
     * @param context an instance of the Context object initialized to a verity-application agent
     * @param reason an human readable reason for the rejection
     * @throws IOException when the HTTP library fails to post to the agency endpoint
     * @throws VerityException when wallet operations fails or given invalid context
     */
    void reject(Context context, String reason) throws IOException, VerityException;

    /**
     * Creates the control message without packaging and sending it.
     * @param context an instance of the Context object initialized to a verity-application agent
     * @return the constructed message (JSON object)
     * @throws VerityException when given invalid context
     *
     * @see #reject
     */
    JSONObject rejectMsg(Context context, String reason) throws VerityException, IOException;

    /**
     * Creates and packages message without sending it.
     * @param context an instance of the Context object initialized to a verity-application agent
     * @return the byte array ready for transport
     * @throws VerityException when wallet operations fails or given invalid context
     *
     * @see #reject
     */
    byte[] rejectMsgPacked(Context context, String reason) throws VerityException, IOException;

    /**
     * Ask for status from the verity-application agent
     *
     * @param context an instance of the Context object initialized to a verity-application agent
     * @throws IOException when the HTTP library fails to post to the agency endpoint
     * @throws VerityException when wallet operations fails or given invalid context
     */
    void status(Context context) throws IOException, VerityException;

    /**
     * Creates the control message without packaging and sending it.
     * @param context an instance of the Context object initialized to a verity-application agent
     * @return the constructed message (JSON object)
     * @throws VerityException when given invalid context
     *
     * @see #status
     */
    JSONObject statusMsg(Context context) throws VerityException;

    /**
     * Creates and packages message without sending it.
     * @param context an instance of the Context object initialized to a verity-application agent
     * @return the byte array ready for transport
     * @throws VerityException when wallet operations fails or given invalid context
     *
     * @see #status
     */
    byte[] statusMsgPacked(Context context) throws VerityException;
}

File: sdk/dotnet-sdk/src/VeritySDK/Protocols/PresentProof/v1_0/PresentProofV1_0.cs
using System;
using System.Json;
using VeritySDK.Protocols;
using VeritySDK.Utils;

namespace VeritySDK.Protocols.PresentProof
{
    /// <summary>
    /// A class for controlling a 1.0 PresentProof protocol.
    /// </summary>
    public class PresentProofV1_0 : AbstractProtocol
    {
        #region Protocol identificator

        /// <summary>
        /// The qualifier for the message family. Uses the community qualifier.
        /// </summary>
        public override string qualifier() { return Util.COMMUNITY_MSG_QUALIFIER; }

        /// <summary>
        /// The name for the message family.
        /// </summary>
        public override string family() { return "present-proof"; }

        /// <summary>
        /// The version for the message family.
        /// </summary>
        public override string version() { return "1.0"; }

        #endregion 

        #region Constructors

        /// <summary>
        /// Constructor
        /// </summary>
        public PresentProofV1_0() { }

        /// <summary>
        /// Constructor
        /// </summary>
        public PresentProofV1_0(string threadId) : base(threadId) { }

        /// <summary>
        /// Constructor
        /// </summary>
        public PresentProofV1_0(string forRelationship, string threadId) : base(threadId)
        {
            this.forRelationship = forRelationship;
        }

        /// <summary>
        /// Constructor
        /// </summary>
        public PresentProofV1_0(string forRelationship, string name, Attribute[] proofAttrs) : this(forRelationship, name, proofAttrs, null) { }

        /// <summary>
        /// Constructor
        /// </summary>
        public PresentProofV1_0(string forRelationship, string name, Predicate[] proofPredicates) : this(forRelationship, name, null, proofPredicates) { }

        /// <summary>
        /// Constructor
        /// </summary>
        public PresentProofV1_0(string forRelationship, string name, Attribute[] proofAttrs, Predicate[] proofPredicates) : this(forRelationship, name, proofAttrs, proofPredicates, false) { }

        /// <summary>
        /// Constructor
        /// </summary>
        public PresentProofV1_0(string forRelationship, string name, Attribute[] proofAttrs, Predicate[] proofPredicates, bool byInvitation)
        {
            this.forRelationship = forRelationship;
            this.name = name;
            this.proofAttrs = proofAttrs;
            this.proofPredicates = proofPredicates;
            this.byInvitation = byInvitation;
            this.created = true;
        }

        /// <summary>
        /// Constructor
        /// </summary>
        public PresentProofV1_0(string forRelationship, string threadId, ProposedAttribute[] attrs, ProposedPredicate[] predicates) : base(threadId) {
            this.forRelationship = forRelationship;
            this.proposedAttributes = attrs;
            this.proposedPredicates = predicates;
            this.created = true;
        }

        #endregion 

        string PROOF_REQUEST = "request";
        string STATUS = "status";
        string REJECT = "reject";
        string ACCEPT_PROPOSAL = "accept-proposal";
        // string PROTOCOL_INVITATION = "protocol-invitation";

        // flag if this instance started the interaction
        bool created = false;

        string forRelationship;
        string name;
        Attribute[] proofAttrs;
        Predicate[] proofPredicates;
        ProposedAttribute[] proposedAttributes;
        ProposedPredicate[] proposedPredicates;
        bool byInvitation;

        /// <summary>
        /// Creates an attribute restriction for a presentation of proof 
        /// </summary>
        /// <param name="name">the attribute name to apply the restriction</param>
        /// <param name="restrictions">an array restrictions to be used</param>
        /// <returns>An Attribute with the given name and restrictions</returns>
        public static Attribute attribute(string name, params Restriction[] restrictions)
        {
            return new Attribute(name, restrictions);
        }

        /// <summary>
        /// Creates a predicate restriction for a presentation of proof. Indy Anoncreds only supports, so the value should be expressed as a greater than predicate. 
        /// </summary>
        /// <param name="name">the attribute name to apply the restriction</param>
        /// <param name="name">(Optional) predicate type (">=", ">", "<=", "<") defaults to ">=" if not included</param>
        /// <param name="value">the value the attribute must be greater than</param>
        /// <param name="restrictions">an array restrictions to be used</param>
        /// <returns>A Predicate with the given name, value and restrictions</returns>
        public static Predicate predicate(string name,string type, int value, params Restriction[] restrictions)
        {
            return new Predicate(name, type, value, restrictions);
        }

        public static Predicate predicate(string name, int value, params Restriction[] restrictions)
        {
            return new Predicate(name, value, restrictions);
        }

        /// <summary>
        /// Directs verity-application to request a presentation of proof. 
        /// </summary>
        /// <param name="context">an instance of the Context object initialized to a verity-application agent</param>
        public void request(Context context)
        {
            send(context, requestMsg(context));
        }

        /// <summary>
        /// Creates the control message without packaging and sending it. 
        /// </summary>
        /// <param name="context">an instance of the Context object initialized to a verity-application agent</param>
        /// <returns>the constructed message (JSON object)</returns>
        public JsonObject requestMsg(Context context)
        {
            if (!created)
            {
                throw new ArgumentException("Unable to request presentation when NOT starting the interaction");
            }

            JsonObject msg = new JsonObject();
            msg.Add("@type", messageType(PROOF_REQUEST));
            msg.Add("@id", getNewId());
            addThread(msg);
            msg.Add("~for_relationship", this.forRelationship);
            msg.Add("name", this.name);
            msg.Add("proof_attrs", JsonUtil.makeArray(proofAttrs));
            if (this.proofPredicates != null)
                msg.Add("proof_predicates", JsonUtil.makeArray(proofPredicates));
            msg.Add("by_invitation", this.byInvitation);
            return msg;
        }

        /// <summary>
        /// Creates and packages message without sending it. 
        /// </summary>
        /// <param name="context">an instance of the Context object initialized to a verity-application agent</param>
        /// <returns>the byte array ready for transport</returns>
        public byte[] requestMsgPacked(Context context)
        {
            return packMsg(context, requestMsg(context));
        }

        /// <summary>
        /// Directs verity-application to accept the request to present proof.
        /// </summary>
        /// <param name="context">an instance of the Context object initialized to a verity-application agent</param>
        public void accept(Context context)
        {
            throw new NotSupportedException();
        }

        /// <summary>
        /// Creates the control message without packaging and sending it. 
        /// </summary>
        /// <param name="context">an instance of the Context object initialized to a verity-application agent</param>
        /// <returns>the constructed message (JSON object)</returns>
        public JsonObject acceptMsg(Context context)
        {
            throw new NotSupportedException();
        }

        /// <summary>
        /// Creates and packages message without sending it. 
        /// </summary>
        /// <param name="context">an instance of the Context object initialized to a verity-application agent</param>
        /// <returns>the byte array ready for transport</returns>
        public byte[] acceptMsgPacked(Context context)
        {
            throw new NotSupportedException();
        }

        /// <summary>
        /// Directs verity-application to accept the proposal for present proof.
        /// Verity-application will send presentation request based on the proposal.
        /// </summary>
        /// <param name="context">an instance of the Context object initialized to a verity-application agent</param>
        public void acceptProposal(Context context)
        {
            send(context, acceptProposalMsg(context));
        }

        /// <summary>
        /// Creates the control message without packaging and sending it.
        /// </summary>
        /// <param name="context">an instance of the Context object initialized to a verity-application agent</param>
        /// <returns>the constructed message (JSON object)</returns>
        public JsonObject acceptProposalMsg(Context context)
        {
            JsonObject msg = new JsonObject();
            msg.Add("@type", messageType(ACCEPT_PROPOSAL));
            msg.Add("@id", getNewId());
            addThread(msg);
            msg.Add("~for_relationship", this.forRelationship);
            return msg;
        }

        /// <summary>
        /// Creates and packages message without sending it.
        /// </summary>
        /// <param name="context">an instance of the Context object initialized to a verity-application agent</param>
        /// <returns>the byte array ready for transport</returns>
        public byte[] acceptProposalMsgPacked(Context context)
        {
            return packMsg(context, acceptProposalMsg(context));
        }

        /// <summary>
        /// Directs verity-application to reject this presentation proof protocol 
        /// </summary>
        /// <param name="context">an instance of the Context object initialized to a verity-application agent</param>
        /// <param name="reason">an human readable reason for the rejection</param>
        public void reject(Context context, string reason)
        {
            send(context, rejectMsg(context, reason));
        }

        /// <summary>
        /// Creates the control message without packaging and sending it. 
        /// </summary>
        /// <param name="context">an instance of the Context object initialized to a verity-application agent</param>
        /// <param name="reason">an human readable reason for the rejection</param>
        /// <returns>the constructed message (JSON object)</returns>
        public JsonObject rejectMsg(Context context, string reason)
        {
            JsonObject msg = new JsonObject();
            msg.Add("@type", messageType(REJECT));
            msg.Add("@id", getNewId());
            addThread(msg);
            msg.Add("~for_relationship", this.forRelationship);
            if (!String.IsNullOrWhiteSpace(reason))
                msg.Add("reason", reason);
            return msg;
        }

        /// <summary>
        /// Creates and packages message without sending it. 
        /// </summary>
        /// <param name="context">an instance of the Context object initialized to a verity-application agent</param>
        /// <param name="reason">an human readable reason for the rejection</param>
        /// <returns>the byte array ready for transport</returns>
        public byte[] rejectMsgPacked(Context context, string reason)
        {
            return packMsg(context, rejectMsg(context, reason));
        }

        /// <summary>
        /// Ask for status from the verity-application agent 
        /// </summary>
        /// <param name="context">an instance of the Context object initialized to a verity-application agent</param>
        public void status(Context context)
        {
            send(context, statusMsg(context));
        }

        /// <summary>
        /// Creates the control message without packaging and sending it.
        /// </summary>
        /// <param name="context">an instance of the Context object initialized to a verity-application agent</param>
        /// <returns>the constructed message (JSON object)</returns>
        public JsonObject statusMsg(Context context)
        {
            JsonObject msg = new JsonObject();
            msg.Add("@type", messageType(STATUS));
            msg.Add("@id", getNewId());
            addThread(msg);
            msg.Add("~for_relationship", this.forRelationship);
            return msg;
        }

        /// <summary>
        /// Creates and packages message without sending it. 
        /// </summary>
        /// <param name="context">an instance of the Context object initialized to a verity-application agent</param>
        /// <returns>the byte array ready for transport</returns>
        public byte[] statusMsgPacked(Context context)
        {
            return packMsg(context, statusMsg(context));
        }
    }
}